### ISSUE TITLE
Can we allow `mixin on <SealedClass>`?

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/language#3215]
There might be a better way to do this, if so please let me know :).

Consider a sealed class hierarchy, with private *Impl classes that expose some shared member only from the *Impls. The *Impl classes also implement another shared interface.

Internally to my code, I want to only deal in the *Impl classes, but I can't declare a type which enforces something to be both the sealed class plus the additional private interface. I end up having to choose one or the other.

### my_public_library.dart

```dart
sealed class Choice {}

abstract interface class ChoiceA implements Choice {}

abstract interface class ChoiceB implements Choice {}
```

### my_private_library.dart

```dart
import 'my_public_library.dart';

class WithId {
  final int id;
  
  WithId(this.id);
}

/// Not allowed, but **might** be safe on its own?
///
/// I believe you would only be able to actually mix this into one of the impl classes.
mixin ChoiceImpl on Choice implements WithId  {}

class ChoiceAImpl extends WithId implements ChoiceA with ChoiceImpl {}

class ChoiceBImpl extends WithId implements ChoiceB with ChoiceImpl {}
```

### usage.dart

```dart
import 'my_private_library.dart';

class ThingWithChoice {
  /// This gives me access to both the `Choice` and `WithId` interfaces, while maintaining
  /// the sealed nature of `Choice` and hiding the `WithId` iterface from users.
  final ChoiceImpl choice;
}
```

**Comment by lrhn** [comment]
This was part of one of my designs along the way. 
The trick is to require the mixin to be `base`-or-more and not allow ignoring that even inside the same library. Same thing we do when extending a `base` class from another library, so nothing new.

Then we can ensure that no _instance_ of a subtype of a sealed class is not also an instance of one of the known direct subclasses, because the point way to create an instance implementering the mixin type, is to apply it to a real subclass of the on-type.

**Comment by jakemac53** [comment]
> The trick is to require the mixin to be `base`-or-more and not allow ignoring that even inside the same library. Same thing we do when extending a `base` class from another library, so nothing new.

Yeah, I was discussing this at lunch with others and the idea of using `base` comes up but of course then I realized you can violate that within your own library today 😭 .

**Comment by jakemac53** [comment]
> The trick is to require the mixin to be `base`-or-more and not allow ignoring that even inside the same library. Same thing we do when extending a `base` class from another library, so nothing new.

Would it be reasonable to do this? Essentially say that you can't violate the normal rules for a `base mixin` if the `on` type is a sealed type from another library? That sounds a bit complicated/special case for this but it would enable what I want to do here.

**Comment by lrhn** [comment]
I already specified it, and fully believe it should be sound and safe.
We decided not to allow it because it was considered too confusing to allow a direct subtype of a sealed class in another library, not because it wasn't safe.

We would have to prevent everyone from implementing the mixin, in the same library or other libraries, even if the mixin is declared in a non-feature library. That's extra special, because they cannot write `base` on the mixin declaration.
The `base` is for your protection, and documentation, the constraint applies anyway.
Also transitively to mixins declared `on` the first mixin.

So it does get complicated.

**Comment by jakemac53** [comment]
Ok, the juice might not be worth the squeeze then in this case. It does feel like a workaround anyways, I was probably trying to be a bit too fancy.

What I _really_ want here is intersection types anyways :).
